src/Forms/FieldList.php

```diff
@@ -59,28 +59,86 @@ public function __clone()
     }
 
     /**
+     * Return an array of all fields, including deeply nested fields.
+     *
+     * @var FieldList
+     */
+    public function nestedFields()
+    {
+        $result = array();
+        $stack = $this->toArray();
+        while ($stack) {
+            /** @var FormField $field */
+            $field = array_shift($stack);
+            $result[] = $field;
+            if ($field instanceof CompositeField) {
+                $stack = array_merge($field->getChildren()->toArray(), $stack);
+            }
+        }
+        return new FieldList($result);
+    }
+
+    /**
      * Return a sequential set of all fields that have data.  This excludes wrapper composite fields
      * as well as heading / help text fields.
      *
      * @return FormField[]
      */
     public function dataFields()
     {
-        if (!$this->sequentialSet) {
-            $this->collateDataFields($this->sequentialSet);
+        if ($this->sequentialSet) {
+            return $this->sequentialSet;
+        }
+        $fields = array();
+        /** @var FormField $field */
+        foreach ($this->nestedFields() as $field) {
+            if (!$field->hasData()) {
+                continue;
+            }
+            $name = $field->getName();
+            if (isset($list[$name])) {
+                $this->fieldNameError($field, __FUNCTION__);
+            }
+            $fields[$name] = $field;
         }
-        return $this->sequentialSet;
+        return $this->sequentialSet = $fields;
     }
 
     /**
      * @return FormField[]
      */
     public function saveableFields()
     {
-        if (!$this->sequentialSaveableSet) {
-            $this->collateDataFields($this->sequentialSaveableSet, true);
+        if ($this->sequentialSaveableSet) {
+            return $this->sequentialSaveableSet;
+        }
+
+        $fields = array();
+        /** @var FormField $field */
+        foreach ($this->nestedFields() as $field) {
+            if (!$field->canSubmitValue()) {
+                continue;
+            }
+            $name = $field->getName();
+            if (isset($list[$name])) {
+                $this->fieldNameError($field, __FUNCTION__);
+            }
+            $fields[$name] = $field;
+        }
+        return $this->sequentialSaveableSet = $fields;
+    }
+    
+    protected function fieldNameError(FormField $field, $functionName)
+    {
+        if ($this->form) {
+            $errSuffix = " in your '{$this->form->class}' form called '" . $this->form->Name() . "'";
```

Comment:
Please use one form of injecting variables into strings. Either concatenate or variable expansion, but not both.